conformance/binary_json_conformance_suite.cc

```diff
@@ -2926,6 +2926,106 @@ void BinaryAndJsonConformanceSuite::RunJsonTestsForFieldMask() {
                                 R"(optional_field_mask: {paths: "foo_3_bar"})");
   ExpectSerializeFailureForJson("FieldMaskTooManyUnderscore", RECOMMENDED,
                                 R"(optional_field_mask: {paths: "foo__bar"})");
+
+  // https://google.aip.dev/161#map-fields introduces map keys as part of FieldMask path
+  // elements. If map keys contain characters otherwise problematic for FieldMask,
+  // backticks can be used for quoting. Backtick itself is escaped using
+  // double-backtick.
+  // If a path element is quoted, no camelCase conversion will be done when serializing
+  // to JSON, it will be copied to JSON as-is, including the quotes.
+  RunValidJsonTest("FieldMaskWithBacktickQuotes", RECOMMENDED,
+                   R"json({"optionalFieldMask": "`foo`"})json",
+                   R"(optional_field_mask: {paths: "`foo`"})");
+
+  RunValidJsonTest("FieldMaskWithBacktickQuotesNoCamelCase", RECOMMENDED,
+                   R"json({"optionalFieldMask": "`foo_bar`"})json",
+                   R"(optional_field_mask: {paths: "`foo_bar`"})");
+
+  RunValidJsonTest("FieldMaskWithBacktickQuotesTwoPaths", RECOMMENDED,
+                   R"json({"optionalFieldMask": "`foo`,`bar`"})json",
+                   R"(optional_field_mask: {paths: "`foo`" paths: "`bar`"})");
+
+  RunValidJsonTest("FieldMaskWithBacktickQuotesTwoElements", RECOMMENDED,
+                   R"json({"optionalFieldMask": "`foo`.`bar`"})json",
+                   R"(optional_field_mask: {paths: "`foo`.`bar`"})");
```

Comment:
Should backtick quoted parts of a path be restricted to only occurring in odd numbered (zero-based) path segments/elements? The above could never be a valid field mask for any message since you cannot have a map of maps (e.g. `map<sting, map<string, string>>` isn't allowed).

Reply:
Great callout, thank you! I fully agree. I fixed the unit tests so that a valid test never uses two consecutive map keys.

Reply:
It might be wise to also include a few parse failure tests for these cases. Failure if the path starts with a quoted segment, failure if there are two quoted segments in a row.

Reply:
I included such cases as well!

I was not 100% confident that a case where a path starts with a quoted segment should be a serialization failure. I think we can confidently say that that is a failure only when we know that a field mask is applied to a regular message. If a field mask for any reason wants to get applied on a map, it might be hurtful to fail it on serialization layer.

TLDR; I agree it's a failure, just not sure on which layer that failure should be surfaced. I added the tests though.

Reply:
Agreed. One problem I find with AIP-161 in general is that it is effectively incompatible with existing implementations. AFAIK every existing implementation validates/parses the JSON representation by:

1. splitting on `,`
2. throwing on any encountered `_`
3. Replace every capital letter with `_` plus lowercase letter.

This string splitting strategy fundamentally cannot work with AIP-161. To be fully conformant the implementation needs to write a parser for reading JSON (throwing during JSON parsing), and a separate parser for any or all of these (throwing during serialization):

1. Programmatically creating a new FieldMask instance from a list/array of paths.
2. Adding a new string path to an existing FieldMask instance
3. Serialization of a FieldMask.

Not to mention any of the FieldMask utilities which might (depending on the implementation) require one to parse individual paths into a "tree" of path segments/elements.

The main point being that all existing implementations that would receive an AIP-161 FieldMask (with backticks and underscores and commas inside the backticks) would fail at parsing the JSON. Not to mention that all the FieldMask utilities would also be non-AIP-161-conforming.

Reply:
@antongrbin , @jcready , @jskeet i went through your comments in this PR and this issue: https://github.com/protocolbuffers/protobuf/issues/12524
I am unable to do apply field mask to map example given in https://google.aip.dev/161, am i missing anything ?
i have tried all the patterns mentioned in there. Is it safe to assume that field mask cannot be applied to keys in a map or struct ?

Reply:
@antongrbin , @jcready , @jskeet i went through your comments in this PR and this issue: #12524
I am unable to do apply field mask to map example given in https://google.aip.dev/161, am i missing anything ?
i have tried all the patterns mentioned in there. Is it safe to assume that field mask cannot be applied to keys in a map or struct ?